tensorboard/uploader/auth.py

```diff
@@ -149,36 +156,147 @@ def clear(self):
+            "url: {url}\n"
+            "code: {code}\n".format(
+                url=device_response["verification_url"],
+                code=device_response["user_code"],
+            )
+        )
+        print(prompt_message)
 
-class CustomInstalledAppFlow(google_auth_oauthlib.flow.InstalledAppFlow):
-    """Customized version of the Installed App OAuth2 flow."""
+        auth_response = self._poll_for_auth_token(
+            device_code=device_response["device_code"],
+            polling_interval=device_response["interval"],
+            expiration_seconds=device_response["expires_in"],
+        )
+
+        return self._build_credentials(auth_response)
 
-    def run(self, force_console=False):
-        """Run the flow using a local server if possible, otherwise the
-        console."""
-        # TODO(b/141721828): make auto-detection smarter, especially for macOS.
-        if not force_console and os.getenv("DISPLAY"):
-            try:
-                return self.run_local_server(port=0)
-            except webbrowser.Error:
-                sys.stderr.write(
-                    "Falling back to console authentication flow...\n"
+    def _send_device_auth_request(self):
+        device_uri = self._client_config["device_uri"]
+        params = {
+            "client_id": self._client_config["client_id"],
+            "scope": " ".join(self._scopes),
+        }
+        r = requests.post(device_uri, data=params).json()
+        if "device_code" not in r:
+            raise RuntimeError(
+                "Auth service temporarily unavailable, try again later."
+            )
+        return r
+
+    def _poll_for_auth_token(
+        self, device_code: str, polling_interval: int, expiration_seconds: int
+    ):
+        token_uri = self._client_config["token_uri"]
+        params = {
+            "client_id": self._client_config["client_id"],
+            "client_secret": self._client_config["client_secret"],
+            "device_code": device_code,
+            "grant_type": self._client_config["grant_type"],
+        }
+        expiration_time = time.time() + expiration_seconds
+        # Error cases documented in
+        # https://developers.google.com/identity/protocols/oauth2/limited-input-device#step-2:-handle-the-authorization-server-response
+        while time.time() < expiration_time:
+            resp = requests.post(token_uri, data=params)
+            r = resp.json()
+            if "access_token" in r:
+                return r
+            if "error" in r and r["error"] == "access_denied":
+                raise PermissionError("Auth was denied.")
+            if resp.status_code in {400, 401}:
+                raise ValueError("There must be an error in the request.")
+
+            if "error" in r and r["error"] == "authorization_pending":
+                time.sleep(polling_interval)
+            elif "error" in r and r["error"] == "slow_down":
+                time.sleep(int(polling_interval * 1.5))
```

Comment:
Added a comment. I don't really expect this to be present in reality, but since it's a retryable error, I opted to handle it somehow.

I was imagining some strange edge-case in which the subsequent request is sent "too quickly" after the interval, and for some reason the server says we're polling too fast. Perhaps this is unrealistic. It shouldn't happen.